### PR TITLE
Feature: Multiple credential types

### PR DESCRIPTION
**Pull request by chriskilding**
Support multiple credential types.

Implements [JENKINS-58339](https://issues.jenkins-ci.org/browse/JENKINS-58339)

## Goals

Credential types:

- [x] SSH Private Key (PEM format)
  - [x] PKCS#1 encoding
  - [x] PKCS#8 encoding
  - [x] OpenSSH encoding
- [x] Certificate (PKCS#12 format)
- [x] Username with Password
- [x] Secret Text

Binding mechanisms:

- Manual job configurator (Web UI):
  - [x] Secret Text
  - [x] Username with Password
  - [x] SSH Private Key
  - [x] Certificate
- `withCredentials` (scripted pipeline DSL):
  - [x] Secret Text
  - [x] Username with Password
  - [x] SSH Private Key
  - [x] Certificate
- `environment` (declarative pipeline DSL):
  - [x] Secret Text
  - [x] Username with Password
  - [x] SSH Private Key

## Non-goals

- Support for proprietary third party credential types, like AWS access keys. (They may be added in future, once the standard credential types have given us a template for how to store and retrieve multiple credential types.)
- Support for `environment` bindings that Declarative Pipeline itself does not support. (Declarative Pipeline does not currently support certificate bindings.)

## Bugs

Showstoppers that have been detected within this PR.

- ~Name fields of AWS Secrets Manager Secrets are not distinct~
- ~Not all credential types bind properly in Declarative Pipeline (`environment` step)~
- ~Credentials not working with git plugin (on agent nodes)~

### CODE REVIEW

**Comment by chriskilding** [comment]
Noteworthy discoveries about credential classes that implement multiple credential interfaces:

- Jenkins credentials list view just presents them as having a single type of its choosing. For example it showed me that my multi-type credential instances were string credentials. This is a harmless cosmetic bug that could be fixed upstream by modifying the UI credential type detector.
- Jenkins credentials binding plugin allows us to bind a multi-type credential to any compatible type. For example it allowed me to bind my multi-type credential either as a Secret text credential (bound to “Variable”) or as a Username and Password (separated or conjoined) credential (bound to “Username Variable” and “Password Variable”).

This provides the groundwork we need to implement the feature.

**Comment by chriskilding** [comment]
A candidate implementation is now complete.

The following questions would need to be answered before proceeding:

- Is the metadata schema following the right naming conventions for AWS tags? (Currently the plugin looks for lower-case bare keys in the secret’s tags which correspond to the credential’s getter method, like “username”. Should it be “Username”? Should it be prefixed with something?)
- Are the examples clear enough in the README?

**Comment by mathewpeterson** [comment]
I definitely agree that the aws tags should have some sort of vendor prefix in order to avoid collision.

**Comment by chriskilding** [comment]
I have an idea what might cause that, and it’s installation-specific.

If the plugin fails to authenticate with AWS (eg wrong AWS access key, wrong IAM permissions), the end result is that the plugin has to throw an unchecked exception when you ask it to list credentials, to say that it could not retrieve them.

Unfortunately it seems that due to the way the rest of Jenkins interacts with the Credentials Provider API, that exception propagates all the way to the top - probably beyond where it should.

As a result, when you view a page containing any credentials from the offending provider, the exception stops Jenkins fetching credentials from other providers too, leading to a blank credentials list even when you know that there should be entries there.

So my first guess is that on your system, the plugin had a wobble when authenticating with AWS. Try some things around that and see if the error still happens.

**Comment by chriskilding** [comment]
Nice work, I'll try to reproduce it locally when I next get the opportunity.

In the meantime, the line throwing the NPE is getting the tags from a `SecretListEntry` and turning them into a `Map<String, String>`. So the tags may have ended up as null in the `SecretListEntry` (or perhaps one of the keys was null, or perhaps a value for a key was null)...

```java
Map<String, String> tags = s.getTags().stream().collect(Collectors.toMap(Tag::getKey, Tag::getValue));
```

**Comment by chriskilding** [comment]
Ah right, so it is possible to end up with a ‘null’ tags property after all. I’ll update the code to defend against this possibility. (I also added defensive checks against tags where the key is null or the value is null, which could also have resulted in NPEs).

**Comment by chriskilding** [comment]
I’ve pushed a fix that should handle your edge case, and a corresponding test for it. Unfortunately Moto (the AWS mock) has a defensive initialiser which turns a null tags list into an empty list in its CreateSecret API handler, so my Moto-based integration tests cannot emulate this condition as they stand.

**Comment by chriskilding** [comment]
Hi Nat,

Could you paste the example keys that didn't work in a comment (I am assuming that they were fake test data - don't paste any 'live' keys in here of course)? I'll write some test cases for them.

If you're interested, you can see the test data I've been using to check the plugin's functionality in `PluginIT`. I thought I had used the output of the commands in the README in those tests, but I could have transcribed the wrong things.

**Comment by chriskilding** [comment]
Hmm, it looks like OpenSSH (the thing underpinning ssh-keygen) changed its default output format last year. It used to output private keys that looked like standard PEM/PKCS1. But nowadays it uses its own proprietary OpenSSH private key format. So the standard BouncyCastle PEMParser rejected it.

IIRC the new OpenSSH key format does at least conform to the ASN.1 spec, so potentially we could try and parse it with the BouncyCastle ASN.1 reader classes.

**Comment by chriskilding** [comment]
On further investigation, the general approach to the new OpenSSH private key format on the Web is... not to even bother.

- If you have an OpenSSH private key already, use a converter command to turn it into a PEM key.
- If you are starting from scratch, use `ssh-keygen -m pem ...` to generate your SSH key in the old PEM format.

**Comment by chriskilding** [comment]
After some exploration of the low level BC classes I’ve been able to add basic support for OpenSSH format private keys.

Due to the limits of what Bouncy Castle supports, for now the plugin will parse the key into a PemObject structure, then test whether the “OPENSSH PRIVATE KEY” header is present, and that the key actually contains some content bytes.

@nathanieltalbot could you give the plugin a go and just confirm that it works with the private key that you were trying to get it to read earlier?

**Comment by chriskilding** [comment]
Regarding the vendor prefix for the metadata keys, I've investigated AWS tagging conventions and come to the following conclusions...

## What's out there today

There are 2 broad types of tags, vendor tags (e.g. AWS-generated tags) and User tags (ones you specify).

**Vendor tags** are typically in URN-like (in Amazon's case, ARN) format. Examples: `aws:autoscaling:groupName`, `aws:createdBy`.

**User tags** are typically bare (i.e. un-prefixed) and have no other convention. Examples I've seen: `Stack`, `Cost Center`, `SquadName`, `PartnerContact`, `name`.

I propose that for AWS tagging purposes, Jenkins is essentially another 'vendor' and therefore deserves a prefix. We can't reasonably have the plugin depend on bare tags like `Username` because, as others have pointed out, the user might already be using them for something else.

## Constraints

- The convention for vendor tags is essentially whatever AWS does, so our name format should be like the AWS format for consistency.
- In case the Jenkins project wants to use URN-like identifiers elsewhere, the prefix should not be `jenkins` alone, or other potential Jenkins properties in this namespace could clash.
- The format should let us specify generally-used metadata keys in a shared way. So that if multiple credential types use a common metadata field like 'username' they should not have to redefine it.
- The format should also allow vendor-specific credential types to define their own metadata keys under a plugin-specific namespace prefix.

## Solution

A URN-like format closely mimicking the AWS one, with a root of `jenkins:credentials`.

Allows common metadata keys to be shared, while individual plugins can also declare more specific keys using the optional `[:prefix]`.

```
jenkins:credentials[:prefix]:<key>
```

Examples:

```
jenkins:credentials:username
jenkins:credentials:aws:accessKeyId
jenkins:credentials:pushbullet:apiKey
jenkins:credentials:azure:<whatever>
```

**Comment by chriskilding** [comment]
@nathanieltalbot The commit I just pushed should ensure that the credential ID is always used for its name. Try it out and see if it fixes the issue for you.

(At this time Moto ignores the input value for a secret’s description, so while I could in principle use the description in the name if it was present, in practice I can’t demonstrate that working with tests... so I will just use the secret ID for now. Hopefully Moto will gain support for secret descriptions in future.)

**Comment by chriskilding** [comment]
I’m now reasonably happy with what the feature looks like and how it would work. Next I will put it through a round of testing in a real multi-node Jenkins setup, to catch any obvious bugs arising from real-world operation.

**Comment by chriskilding** [comment]
Sure enough, a pair of bugs popped up :) One is an easy fix, the second is harder...

When I try to bind Secrets Manager credentials in a declarative Jenkinsfile using EnvInject, i.e. `environment { FOO = credentials(‘foo’) }`, something tries to bind all types of credentials as SSH keys. I recompiled the plugin without SSH key support and the EnvInject binder then behaved properly.

So I have two questions for the Jenkins project veterans on this thread:

1. Where does the credential -> environment variable binding code live? (I can’t find it.)
2. Why would the EnvInject binder have a particular fixation on SSH keys? Does it have special-case handling for them?

**Comment by coltrey** [comment]
I'm definitely not the veteran on this topic, but was curious about a couple of things:

> When I try to bind Secrets Manager credentials in a declarative Jenkinsfile using EnvInject, i.e. `environment { FOO = credentials(‘foo’) }`, something tries to bind all types of credentials as SSH keys. I recompiled the plugin without SSH key support and the EnvInject binder then behaved properly.

Looking at this syntax - how does the binder divine what the credential type is?  In the scripted pipeline syntax, the credential type is explicitly specified, but in the declarative syntax, the type is not explicitly specified, so it's unclear to me how the binder ever determines that.

Looking at the AWSCredentials class, I couldn't answer this question: how does the AWSCredentials class tell its callers what type of credential it has implemented?

> 1. Where does the credential -> environment variable binding code live? (I can’t find it.)

Have you already looked [here](https://github.com/jenkinsci/credentials-binding-plugin/blob/2a0d796a742ea089fcebfa1d8170326b420fbfe5/src/main/java/org/jenkinsci/plugins/credentialsbinding/)?

**Comment by chriskilding** [comment]
Hi Coltrey,

Yes, I was looking exactly for the code where the binder ‘divines’ the type of the credential.

I’d looked at credentials-binding-plugin, envinject-plugin, and credentials-plugin and I had seen this:

```
public abstract class MultiBinding ... {

  abstract Class<C> type();

}
```

and its subclasses `StringBinding`, `UsernamePasswordBinding`, `SSHUserPrivateKeyBinding` (each returns their respective type class in that method).

However the only instantiation points I can see for those classes are the `@DataBoundConstructor`s, which correspond to the `withCredentials` DSL step. I cannot find the binding code corresponding to the `environment` DSL step - if you have any ideas where that is, then that would be super helpful.

**Comment by coltrey** [comment]
> Hi Coltrey,
> 
> Yes, I was looking exactly for the code where the binder ‘divines’ the type of the credential.
> 
> I’d looked at credentials-binding-plugin, envinject-plugin, and credentials-plugin and I had seen 

Perhaps something in the [model definition plugin](https://github.com/jenkinsci/pipeline-model-definition-plugin/blob/7502d14aa4db891302abf48a2c7de7c9ced97f95/DEV_GUIDE.md#environment-resolution) determines how it gets resolved...

**Comment by coltrey** [comment]
> Specifically, [this thing.](https://github.com/jenkinsci/pipeline-model-definition-plugin/blob/master/pipeline-model-extensions/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/model/CredentialsBindingHandler.java#L90)

Seems to me like the forCredential method in the CredentialsBindingHandler class iterates over the list of credential binding handlers and selects the first one that claims to work.  What that means, I think, is happening is that they're probably roughly alphabetical, so FileCredentialsHandler says it can't handle the AWSCredentials, then moves to SSHUserPrivateKeyHandler, which can.  When you remove the SSHCredential interface from the AWSCredentials class, that handler now cannot, so it moves on to the next one - StringHandler.  I bet this bug would exist in that scenario if you were trying a credential that was supposed to use the  UsernamePasswordHandler - I think in that case, the StringHandler would be used because AWSCredentials implements the StringCredentials interface and that handler is alphabetically before the UsernamePasswordHandler. (if alphabetism isn't the key thing here, that behavior would be reversed).

**Comment by chriskilding** [comment]
I’ve now arranged some integration tests that run small declarative and scripted pipeline examples, to formally demonstrate `withCredentials()` and `environment` bindings working (or not).

I discovered a new chicken-and-egg style bug through this work, which I think would affect *all* remote credentials providers (not just this one)...

I installed a plugin on my test Jenkins that needed a username/password credential set in its global configuration (more specifically, it was the Docker plugin, and it needed a global Docker Registry API key to be set).

Now this causes a chicken-and-egg problem... if the remote credentials provider is not yet set up, when the global configuration page is loaded it dutifully throws an exception, just as the implementor’s guide says it should. This crashes the configuration page. Which means the credentials provider’s settings - if there are any - cannot be configured on that page.

I believe that at the moment, the only way around this would be to configure the installation using Configuration As Code: this avoids touching the configuration page until at least the credentials provider is set up.

Could others confirm whether this is indeed a more generally applicable bug? If so I’ll file it on Jira (but I’m not exactly sure what I’d file it against, since it would affect *all* remote credentials providers, but (probably) not local ones).

**Comment by red8888** [comment]
@chriskilding would this PR allow me to have my own json schema? would I be able to have secrets defined in json in a specific way like this:
```
my-aws-secret value:
{ 
   "user1":"password1",
   "user2":"password2",
   "user3":"password3",
   "serviceblah": { 
      "entity1":"key1",
      "entity2":"key2"
   }
}
```
And then reference them some way like this:
```
environment { FOO = credentials('my-aws-secret:user1:password1') } // get password1
environment { FOO = credentials('my-aws-secret:serviceblah:entity1:key1') } // get nested key1
```

if not is there plans to add something like this? I have applications that expect secrets in a certain format- some of these secrets I'd like to use as Jenkins bindings without having to recreate them in a different format and have to manage two copies

**Comment by chriskilding** [comment]
I’ve now added a custom binding handler to make the plugin work with the declarative pipeline (the `environment {}` workflow).

This handler necessarily depends on types in pipeline-model-definition. I’d therefore like to make it an @OptionalExtension with `variant`, so that plugin users do not have to install pipeline-model-definition unless they actually use Pipelines.

Unfortunately when I changed the handler annotation from @Extension to @OptionalExtension, the Jenkins extension loader ignored the handler and acted like it didn’t exist.

@coltrey and other experts - any ideas what I’m missing?

**Comment by coltrey** [comment]
> @coltrey and other experts - any ideas what I’m missing?

By no means am I an expert, but if you check your code in somewhere I can take a look.

- Are you certain the activation criteria for the variant is being triggered?
- Is there any debug messaging that can be enabled? (It seems in the variant code, the only messaging is a warning when parsing a manifest)

**Comment by chriskilding** [comment]
I have reported the /configure page load order bug here: [JENKINS-59626](https://issues.jenkins-ci.org/browse/JENKINS-59626).

Because this bug broke the plugin's web configuration tests, my only recourse was to disable those tests for the time being, and to create equivalent tests which ensure the same configuration properties can be set via other means (like Configuration As Code and the REST API).

**Comment by chriskilding** [comment]
@red8888 This PR is concerned with the representation of multiple credential types, not the number of credentials per secret, so what you’re asking for would be out of scope of this PR. However I’ll add it as a Jenkins Jira ticket so that it’s remembered, and we can return to it in the future.

**Comment by salemgolemugoo** [comment]
@chriskilding merge?

**Comment by chriskilding** [comment]
Hi @salemgolemugoo - we appreciate that you (like teams in our own company) are hungry for the feature. Me and @chroche are working as fast as we can to stamp out the bugs before I go ahead and make a beta version available for wider testing.

The change is quite substantial, and credential providers sit in a critical place in the Jenkins stack i.e. 
- they **have** to work (or they’ll crash everyone’s Jenkins builds)
- they **have** to be secure (or they’ll leak the very things they exist to protect)
- they **have** to be performant (or they’ll cause job timeouts and UI freezes)

So therefore we have to exercise a lot of caution when updating such a component.

**Comment by chriskilding** [comment]
We think we’re getting close to a beta though - we’ve got to the point where we are mostly tackling bugs that only appear in specific credentials consumer plugins, and in multi-node environments (which can be extremely difficult to write tests for).

**Comment by chriskilding** [comment]
Right, we think we’ve nailed the multi-node bug. Just one question about the solution - what I’ve done is *functional* but I don’t know if it is *correct*.

Currently I have serialised the relevant AWS client config so it can survive transport over to the agent node. Therefore the credential value resolution happens on the agent itself. So the agents must have access to the backing store (Secrets Manager) as well as the master.

Is this what consumers like `git-client-plugin` are expecting to happen when they call `CredentialsProvider.snapshot(cred)`?

Or do they really want to be passed a resolved version of the credential, which contains the secret value stored on the object?

**Comment by coltrey** [comment]
> Running Declarative Pipeline job: O(2B + 1)

If this is worst case, it seems minor performance-wise.  As for the failure points, how do the other external providers handle this? (i.e.: retry option?) I wonder what the [kubernetes provider](https://github.com/jenkinsci/kubernetes-credentials-provider-plugin)does.  If this plugin is similar, seems to me like it would be okay.

**Comment by chriskilding** [comment]
Yeah that was my intuition, the number of calls could be lower, but the real goal was avoiding the O(N) strategies for type detection, which would lead to Jenkins making potentially hundreds or even thousands of calls at once.

Regarding failures, IIRC the AWS SDK has built-in retry with backoff and jitter, for at least some kinds of errors. This is given to us for free and we don’t have to configure anything to use it.

This means the AWS SDK should have a good go at trying to get the secret value before giving up, and we throw the CredentialsUnavailableException.

However it was a while ago that I went source code diving in the AWS SDK so if someone has newer experience then please chime in :)

**Comment by itsSaad** [comment]
Thankyou guys for this amazing feature and the equally amazing hard work. We are definitely waiting for this to be released and will be very excited to try this out.

**Comment by chriskilding** [comment]
In terms of caching I think the plugin is already doing about as much as it can...

It’s caching all invocations of ListSecrets already (5min duration).

The scripted pipeline results in 1 GetSecretValue call per cred, at the start of each build, then caches for the build’s duration.

The declarative pipeline is the same except it’s 2 calls per cred.

If a plugin (eg git-client-plugin) elects to snapshot a credential from secrets manager, a single GetSecretValue call will happen at snapshot time, and then the snapshotted cred will exist as long as that plugin wants it.

This is all pretty efficient... I can’t think of what else I could cache. And to exceed the 700rps limit, I can only think of 2 scenarios:

- A Jenkins installation that is spawning say ~200 pipeline jobs *at the same time* with say 3/4 creds each.
- Credentials consumer plugins that directly call the CredentialsProvider and then handle the returned cred objects in a custom way. (They might keep calling getSecret() and not cache it. We can’t stop them.)

In these scenarios AWS SDK should throw a LimitExceededException. I trap AmazonClientExceptions (the general AWS exception supertype) to throw a more friendly CredentialsUnavailableException for the field that could not be resolved. But I do not currently include the message from the AmazonClientException (as it can be long).

Maybe I should adjust this behaviour to include the AWS message, or a short form of it?

**Comment by coltrey** [comment]
> Maybe I should adjust this behaviour to include the AWS message, or a short form of it?

I'd say include the detailed message at INFO or greater to allow an administrator to create a system log that will catch it for troubleshooting.

**Comment by chriskilding** [comment]
The Username With Password and SSH Key credential types have metadata fields and I believe all of those should reflect the tagging naming conventions. (Unless you can highlight any I’ve missed?)

The Secret Text and Certificate types don’t have any metadata fields.

**Comment by chriskilding** [comment]
I’m pleased to announce that me and @chroche have completed our private testing of this feature, and squashed the initial bugs that we could find.

It’s public review time...

**Comment by chriskilding** [comment]
@itsSaad @nathanieltalbot @jeffpearce I believe each of you have Jenkins installations and an interest in running this plugin - would you be happy to test drive this feature and let me know if it works well for you?

**Comment by jeffpearce** [comment]
I’m happy to give them a try

File: src/main/java/io/jenkins/plugins/credentials/secretsmanager/AwsCredentials.java [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+ * particular credential type. If these things are not present, the relevant accessor method(s) will
+ * fail at lookup time. (For example, to use the AWS secret as a Jenkins {@link SSHUserPrivateKey},
+ * the secretString must be in private key format, and username metadata must be present in the
+ * secret's tags.)
+ */
+@NameWith(value = AwsCredentials.NameProvider.class)
+abstract class AwsCredentials extends BaseStandardCredentials implements StringCredentials, StandardUsernamePasswordCredentials, SSHUserPrivateKey, StandardCertificateCredentials {
+
+    private static final char[] EMPTY_PASSWORD = {};
+    private static final Secret NONE = Secret.fromString("");
```
Comment (jeffpearce) [inline src/main/java/io/jenkins/plugins/credentials/secretsmanager/AwsCredentials.java]: Super nit, but I'd consider calling this NO_SECRET or NONE_SECRET. That would (arguably) increase code readability IMO

File: src/main/java/io/jenkins/plugins/credentials/secretsmanager/AwsCredentials.java [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    private String getSecretString() {
+        return getSecretValue().match(new SecretValue.Matcher<String>() {
+            @Override
+            public String string(String str) {
+                return str;
+            }
+
+            @Override
+            public String binary(byte[] bytes) {
+                return null;
```
Comment (jeffpearce) [inline src/main/java/io/jenkins/plugins/credentials/secretsmanager/AwsCredentials.java]: I'm not sure of the API spec, but is it possible to return something here? If it's not expected to be called, would it make sense to throw an exception?

File: src/main/java/io/jenkins/plugins/credentials/secretsmanager/AwsCredentialsProvider.java [hunk 3]
```diff
@@ -28,17 +49,16 @@
     private final AwsCredentialsStore store = new AwsCredentialsStore(this);
 
     private final Supplier<Collection<IdCredentials>> credentialsSupplier =
-            CredentialsSupplierFactory.create();
+            memoizeWithExpiration(AwsCredentialsProvider::fetchCredentials, Duration.ofMinutes(5));
```
Comment (jeffpearce) [inline src/main/java/io/jenkins/plugins/credentials/secretsmanager/AwsCredentialsProvider.java]: Curious where the 5 minutes value comes from. Would it make sense to make it configurable at some point?

Comment (chriskilding) [reply src/main/java/io/jenkins/plugins/credentials/secretsmanager/AwsCredentialsProvider.java]: 5 minutes is IIRC the recommended default in the Credentials Provider implementor’s guide. Making it user-configurable is indeed on the to do list: [JENKINS-57577](https://issues.jenkins-ci.org/browse/JENKINS-57577)

File: src/main/java/io/jenkins/plugins/credentials/secretsmanager/AwsCredentials.java (hunk 2 above)
Comment (chriskilding) [reply src/main/java/io/jenkins/plugins/credentials/secretsmanager/AwsCredentials.java]: Thinking in terms of K-maps, an AWS Secret that contains neither a string nor a binary value should be a “can’t happen” state: AWS would have slipped up badly on their input validation for that to occur. I guess it would be prudent to throw an IllegalStateException then

Comment (chriskilding) [reply src/main/java/io/jenkins/plugins/credentials/secretsmanager/AwsCredentials.java]: annotated getSecretValue() with `@NonNull`, threw exception if the AWS call ever gets to this state

File: src/main/java/io/jenkins/plugins/credentials/secretsmanager/AwsCredentials.java (hunk 1 above)
Comment (chriskilding) [reply src/main/java/io/jenkins/plugins/credentials/secretsmanager/AwsCredentials.java]: Done

**Comment by chriskilding** [comment]
Hi Nat, I’d be interested specifically if you could test the latest build with a job that has a Git checkout step, or sshagent step, running on a slave node. (Credential snapshotting and serialisation is something I’ve only addressed in the latest builds - but I’d appreciate confirmation of whether it is indeed fixed.)

**Comment by chriskilding** [comment]
And yeah ci.jenkins.io was down for all when I pushed a few hours ago. Some of the project leads are re-prioritising their upcoming work to focus on the infrastructure and outages, which is good news.

**Comment by chriskilding** [comment]
@Quaiks I believe you created the original JIRA ticket for this feature, so would you be able to have a play with it and let me know if it works for your company?

**Comment by chriskilding** [comment]
In the interests of time (and some upcoming Jenkins deployments at work) I intend to merge this in the next couple of weeks.

So if you have any feedback, please post it below before the 22nd November. If I don’t hear anything by then I’ll assume it’s good to go and merge to master.

**Comment by chriskilding** [comment]
Alright, looks like it's merge time...

**Comment by chriskilding** [comment]
I've uploaded a release candidate `0.1.0-beta-1` to Jenkins Experiments. We are smoke testing that now. If all goes to plan the stable release should follow in a few days time.

**Comment by chriskilding** [comment]
Stable version 0.1.0 has been uploaded to the main Jenkins update site. This concludes the feature.

File: src/main/java/io/jenkins/plugins/credentials/secretsmanager/AwsCredentialsSnapshotTaker.java [hunk 4]
```diff
@@ -0,0 +1,20 @@
+package io.jenkins.plugins.credentials.secretsmanager;
+
+import com.cloudbees.plugins.credentials.CredentialsSnapshotTaker;
+
+import hudson.Extension;
+
+@Extension
+@SuppressWarnings("unused")
+public class AwsCredentialsSnapshotTaker extends CredentialsSnapshotTaker<AwsCredentials> {
```
Comment (jglick) [inline src/main/java/io/jenkins/plugins/credentials/secretsmanager/AwsCredentialsSnapshotTaker.java:9]: https://github.com/jenkinsci/azure-keyvault-plugin/pull/96#issuecomment-998119909

I see that your ITs directly call `CredentialsProvider.snapshot`, but that is not really testing whether an agent build would do the same.

Comment (chriskilding) [reply src/main/java/io/jenkins/plugins/credentials/secretsmanager/AwsCredentialsSnapshotTaker.java:9]: I believe that calling snapshot is something the downstream credentials consumer plugin has to do. As previously mentioned this isn't ideal; the main consumer plugins like pipeline do this, but other plugins don't.

Would there be a central higher-level place, maybe in the Credentials API, where we could add something to always snapshot the retrieved credential if a plugin is running in a remote agent context?

Comment (jglick) [reply src/main/java/io/jenkins/plugins/credentials/secretsmanager/AwsCredentialsSnapshotTaker.java:9]: > the main consumer plugins like pipeline do this, but other plugins don't

AFAICT only `git-client` does it, because the weird implementation of `git-client` + `git` sends a remoted proxy object for all operations. (Other SCM plugins get by without such tricks.) I have not come across other plugins which serialize `Credentials` instances over a Remoting channel but it would be hard to search for this pattern.

> we could add something to always snapshot the retrieved credential if a plugin is running in a remote agent context

I suppose `BaseCredentials` could implement `writeReplace` for this purpose (taking care to differentiate Remoting from XStream serialization). I am not sure what such a change could break. I came across a `CredentialsProvider.snapshot` call site in CloudBees CI written by the author of `CredentialsSnapshotTaker` which makes me nervous about the compatibility implications, unless `CredentialsSnapshotTaker` implementations are always idempotent.

Ideally integration tests would exercise authenticated Git checkouts. That could be pretty tricky though. Even assuming e.g. a WireMock for Secrets Manager, `GitSampleRepoRule` does not support authentication.
